src/tweedle/fields/fp.rs

```diff
@@ -569,6 +544,150 @@ impl Field for Fp {
         CtOption::new(tmp, !self.ct_eq(&Self::zero()))
     }
 
+    fn pow_vartime<S: AsRef<[u64]>>(&self, exp: S) -> Self {
+        let mut res = Self::one();
+        let mut found_one = false;
+        for e in exp.as_ref().iter().rev() {
+            for i in (0..64).rev() {
+                if found_one {
+                    res = res.square();
+                }
+
+                if ((*e >> i) & 1) == 1 {
+                    found_one = true;
+                    res *= self;
+                }
+            }
+        }
+        res
+    }
+}
+
+#[cfg(not(target_pointer_width = "64"))]
+type ReprBits = [u32; 8];
+
+#[cfg(target_pointer_width = "64")]
+type ReprBits = [u64; 4];
+
+impl ff::PrimeField for Fp {
+    type Repr = [u8; 32];
+    type ReprBits = ReprBits;
+
+    const NUM_BITS: u32 = 255;
+    const CAPACITY: u32 = 254;
+    const S: u32 = S;
+
+    fn from_repr(repr: Self::Repr) -> Option<Self> {
+        Self::from_bytes(&repr).into()
+    }
+
+    fn to_repr(&self) -> Self::Repr {
+        self.to_bytes()
+    }
+
+    fn to_le_bits(&self) -> BitArray<Lsb0, Self::ReprBits> {
+        let bytes = self.to_bytes();
+
+        #[cfg(not(target_pointer_width = "64"))]
+        let limbs = [
+            u32::from_le_bytes(bytes[0..4].try_into().unwrap()),
+            u32::from_le_bytes(bytes[4..8].try_into().unwrap()),
+            u32::from_le_bytes(bytes[8..12].try_into().unwrap()),
+            u32::from_le_bytes(bytes[12..16].try_into().unwrap()),
+            u32::from_le_bytes(bytes[16..20].try_into().unwrap()),
+            u32::from_le_bytes(bytes[20..24].try_into().unwrap()),
+            u32::from_le_bytes(bytes[24..28].try_into().unwrap()),
+            u32::from_le_bytes(bytes[28..32].try_into().unwrap()),
+        ];
+
+        #[cfg(target_pointer_width = "64")]
+        let limbs = [
+            u64::from_le_bytes(bytes[0..8].try_into().unwrap()),
+            u64::from_le_bytes(bytes[8..16].try_into().unwrap()),
+            u64::from_le_bytes(bytes[16..24].try_into().unwrap()),
+            u64::from_le_bytes(bytes[24..32].try_into().unwrap()),
+        ];
+
+        BitArray::new(limbs)
+    }
+
+    fn is_odd(&self) -> bool {
+        todo!()
```

Comment:
Should this be the usual definition?
```suggestion
        self.to_bytes()[0] & 1 == 1
```

Reply:
Yes.